[RISC-V] coreclr-vm and other directories in coreclr

clamp03 opened the pull request:
- Successfully cross-build for RISC-V.
- Run A simple application "helloworld"
- Fail a test in clr.paltest

clamp03 committed:
[RISC-V] coreclr-vm and other directories in coreclr

- Successfully cross-build for RISC-V.
- Run A simple application "helloworld"
- Fail a test in clr.paltest

clamp03 committed:
Remove commented codes

clamp03 commented:
@mangod9 Could you review this PR? Actually I don't know why tests failed or cancelled.
If you leave any comment, I will update as possible as I can.
Thank you.

mangod9 commented:
@janvorli as well. Also are you asking about why the superpmi tests are failing or something else?

clamp03 commented:
Actually, I am asking reviews from you and owners to merge the PR. To get review, I tried to pass all tests, but some tests failed and I cannot find the reasons. :)

mangod9 commented on src/coreclr/inc/regdisp.h:
```diff
@@ -397,7 +425,24 @@ inline void FillContextPointers(PT_KNONVOLATILE_CONTEXT_POINTERS pCtxPtrs, PT_CO
     }
-#else // TARGET_X86
+#elif defined(TARGET_RISCV64) // TARGET_X86
+    // *(&pCtxPtrs->S0) = &pCtx->S0;
```
is this commented out on purpose?

clamp03 replied:
Actually, I did on purpose. In RISC-V, S0 is used as FP. (FP is handled in line 442.) I want to keep it as commented.
However, I think it can be removed.

mangod9 commented on src/coreclr/inc/volatile.h:
```diff
@@ -68,8 +68,8 @@
-#if defined(__GNUC__) && !defined(HOST_X86) && !defined(HOST_AMD64) && !defined(HOST_ARM) && !defined(HOST_ARM64) && !defined(HOST_LOONGARCH64) && !defined(HOST_RISCV64) && !defined(HOST_S390X) && !defined(HOST_POWERPC64)
-#error The Volatile type is currently only defined for GCC when targeting x86, AMD64, ARM, ARM64, LOONGARCH64, RISCV64, PPC64LE, or S390X CPUs
+#if defined(__GNUC__) && !defined(HOST_X86) && !defined(HOST_AMD64) && !defined(HOST_ARM) && !defined(HOST_ARM64) && !defined(HOST_LOONGARCH64) && !defined(HOST_RISCV64) && !defined(HOST_S390X) && !defined(HOST_POWERPC64) && !defined(HOST_RISCV64)
+#error The Volatile type is currently only defined for GCC when targeting x86, AMD64, ARM, ARM64, LOONGARCH64, RISCV64, PPC64LE, S390X, or RISCV64 CPUs
```
`RISCV64` was already included here?

clamp03 replied:
Oops. Thank you!

mangod9 commented on src/coreclr/vm/codeman.cpp:
```diff
@@ -909,6 +909,29 @@ BOOL IsFunctionFragment(TADDR baseAddress, PTR_RUNTIME_FUNCTION pFunctionEntry)
         pUnwindCodes += EpilogCount;
     }
 
+    return ((*pUnwindCodes & 0xFF) == 0xE5);
```
this looks equivalent to `TARGET_LOONGARCH64` perhaps just keep it a single implementation?

clamp03 replied:
Registers and assembly instructions of RISC-V and LOONGARCH64 are more similar than ARM. I implemented RISC-V based on LOONGARCH64.

mangod9 replied:
hence wondering if they are the same can we remove duplication for now and just use the LoongArch implementation or is it expected to diverge soon?

clamp03 replied:
Sorry. I misunderstood your review. I updated it.
Thank you!

mangod9 commented on src/coreclr/vm/codeman.cpp:
```diff
@@ -1113,6 +1136,44 @@ PTR_VOID GetUnwindDataBlob(TADDR moduleBase, PTR_RUNTIME_FUNCTION pRuntimeFuncti
     *pSize = size;
     return xdata;
 
+#elif defined(TARGET_RISCV64)
```
same as above?

mangod9 commented on src/coreclr/vm/codeman.cpp:
```diff
@@ -2235,6 +2296,8 @@ BOOL EEJitManager::LoadJIT()
 #elif defined(TARGET_LOONGARCH64)
             altJitName = MAKEDLLNAME_W(W("clrjit_unix_loongarch64_loongarch64"));
+#elif defined(TARGET_RISCV64)
+            altJitName = MAKEDLLNAME_W(W("clrjit_unix_riscv64_riscv64"));
```
is riscv64 supported on windows?

clamp03 replied:
No. And I don't have any plan to support on windows now.

mangod9 commented on src/coreclr/vm/fieldmarshaler.h:
```diff
@@ -92,7 +92,20 @@ class NativeFieldDescriptor
     }
 
-    PTR_MethodTable GetNestedNativeMethodTable() const;
+    PTR_MethodTable GetNestedNativeMethodTable() const
```
is this definition moved here for consistency?

clamp03 replied:
This is for resolving a build error.
```
  [ 93%] Building CXX object vm/wks/CMakeFiles/cee_wks_core.dir/__/weakreferencenative.cpp.o
  /usr/bin/riscv64-linux-gnu-ld: ../../dlls/mscordac/libmscordaccore.so: undefined reference to `NativeFieldDescriptor::GetNestedNativeMethodTable() const'
  clang: error: linker command failed with exit code 1 (use -v to see invocation)               
  make[2]: *** [debug/createdump/CMakeFiles/createdump.dir/build.make:308: debug/createdump/createdump] Error 1
```
fieldmarshalre.cpp is not included for createdump.
I resolved this error by putting the implemention into header.

mangod9 commented on src/coreclr/vm/jitinterface.cpp:
```diff
@@ -11246,6 +11254,14 @@ void CEEJitInfo::allocUnwindInfo (
+    ULONG * pPersonalityRoutineRW = (ULONG*)((BYTE *)pUnwindInfoRW + ALIGN_UP(unwindSize, sizeof(ULONG)));
+    *pPersonalityRoutineRW = ExecutionManager::GetCLRPersonalityRoutineValue();
+
+;
```
unnecessary `;`?

clamp03 replied:
Sorry. My mistake.

clamp03 committed:
Fix mistakes

clamp03 committed:
Update by reviews

clamp03 committed:
[VM] Fix test

clamp03 committed:
[VM] Update

Updated by review in #82380

clamp03 committed:
[VM] Update assert and todo comments

clamp03 committed:
[VM] FIX TEST ERRORS

clamp03 committed:
[VM] Update Stubs

gbalykov reviewed (commented):
Quickly looked through half of the files, have some comments

clamp03 committed:
[VM] Updated

clamp03 commented on src/coreclr/gc/env/gcenv.base.h:
```diff
@@ -226,6 +226,11 @@ typedef DWORD (WINAPI *PTHREAD_START_ROUTINE)(void* lpThreadParameter);
 #endif // __loongarch64
 
+#ifdef __riscv
+ #define YieldProcessor() asm volatile( "fence iorw, iorw");
```
Thank you. Fixed.

clamp03 commented on src/coreclr/inc/clrconfigvalues.h:
```diff
@@ -737,12 +742,12 @@ RETAIL_CONFIG_DWORD_INFO(INTERNAL_GDBJitEmitDebugFrame, W("GDBJitEmitDebugFrame"
 //
-#if defined(TARGET_LOONGARCH64)
+#if defined(TARGET_LOONGARCH64) || defined(TARGET_RISCV64)
 //TODO: should implement LoongArch64's features.
```
Thank you. Fixed.

clamp03 commented on src/coreclr/vm/arm64/cgencpu.h:
```diff
@@ -2,7 +2,6 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 //
 
-
```
Thank you. Fixed.

clamp03 commented on src/coreclr/gcinfo/CMakeLists.txt:
```diff
@@ -79,8 +79,12 @@ if (CLR_CMAKE_TARGET_ARCH_LOONGARCH64)
+if (CLR_CMAKE_TARGET_ARCH_RISCV64)
+  create_gcinfo_lib(TARGET gcinfo_unix_riscv64 OS unix ARCH riscv64)
+else()
+  create_gcinfo_lib(TARGET gcinfo_universal_arm OS universal ARCH arm)
```
Thank you. Fixed.

clamp03 commented on src/coreclr/inc/clrconfigvalues.h:
```diff
@@ -293,7 +293,12 @@ CONFIG_DWORD_INFO(INTERNAL_JitDebuggable, W("JitDebuggable"), 0, "")
 RETAIL_CONFIG_DWORD_INFO(INTERNAL_JitEnableNoWayAssert, W("JitEnableNoWayAssert"), INTERNAL_JitEnableNoWayAssert_Default, "")
+
+#if defined(TARGET_RISCV64)
+RETAIL_CONFIG_DWORD_INFO(UNSUPPORTED_JitFramed, W("JitFramed"), 1, "Forces EBP frames")
```
Thank you. Fixed.

clamp03 commented on src/coreclr/inc/cordebuginfo.h:
```diff
@@ -179,6 +179,40 @@ class ICorDebugInfo
+        REGNUM_T0,
+        REGNUM_T1,
+        REGNUM_T2,
+        REGNUM_S0,
```
Right! Fixed.

clamp03 commented on src/coreclr/vm/gccover.h:
```diff
@@ -112,6 +112,12 @@ typedef DPTR(GCCoverageInfo) PTR_GCCoverageInfo; // see code:GCCoverageInfo::sav
 
+#elif defined(TARGET_RISCV64)
+// TODO-RISCV64: Confirm the following encodings are undefined
+#define INTERRUPT_INSTR                 0xBADC0DE0
```
Thank you. I chaged to `0x20000000`, `0x20010000` and `0x20020000`.

t-mustafin commented on src/coreclr/vm/riscv64/asmhelpers.S:
```diff
@@ -1,7 +1,967 @@
+    RestoreRegMS  9, s9
+    RestoreRegMS  10, s10
+    RestoreRegMS  11, s11
+    RestoreRegMS  12, tp
+    RestoreRegMS  13, gp
```
As said in `stubs.cpp` and in comment above `gp` is 12, `sp` is 13.

clamp03 replied:
Thank you. Fixed

t-mustafin commented on src/coreclr/vm/riscv64/asmhelpers.S:
```diff
@@ -1,7 +1,967 @@
+    sd  s9, 72(a1)
+    sd  s10, 80(a1)
+    sd  s11, 88(a1)
+    sd  tp, 96(a1)
+    sd  gp, 104(a1)
```
Same, `gp` <-> `tp`.

clamp03 replied:
Thank you. Fixed

clamp03 committed:
[VM] Fix a bug

t-mustafin committed:
Fix exceptionhandling, stack smashing detected

clamp03 committed:
[VM] Add getRISCV64PassStructInRegisterFlags

clamp03 committed:
Revert "[VM] Add getRISCV64PassStructInRegisterFlags"

This reverts commit cd1ea45de36458a5e3ec66a325bf3e0c4812680d.

clamp03 commented:
Revert `[[VM] Add getRISCV64PassStructInRegisterFlags](https://github.com/dotnet/runtime/pull/82381/commits/cd1ea45de36458a5e3ec66a325bf3e0c4812680d)` and apply to https://github.com/dotnet/runtime/pull/82379 for coherent JIT-EE implementations/consumptions (https://github.com/dotnet/runtime/pull/82379#discussion_r1159548562)

clamp03 committed:
[VM] Restore getLoongArch64PassStructInRegisterFlags

In coreclr-jit patch, it makes getRISCV64PassStructInRegisterFlags for
RISCV64. So restore getLoongArch64PassStructInRegisterFlags

clamp03 commented:
@mangod9 Could you review it? And if it is good, could you please merge it? I fixed all the things requested by the reviewers and waited 50 days. Other two PRs (https://github.com/dotnet/runtime/pull/82379 , https://github.com/dotnet/runtime/pull/82382) are already merged. Actually, I cannot fix all bugs and unimplemented code with just this PR. After the PR is merged, I will continue to fix and my colleagues and community members can help to support RISC-V in coreclr better.

janvorli reviewed (approved):
LGTM, thank you!

clamp03 commented:
> LGTM, thank you!

Thank you for the approval. If it is okay, could you merge this PR? If it needs more approvals or something else, please let me know.

clamp03 committed:
Merge remote-tracking branch 'upstream/main' into merge

clamp03 commented:
> @clamp03 there are some conflicts, can you please resolve them?

Sure. I resolved and pushed a commit. Thank you!

clamp03 committed:
[VM] Fix vm/riscv64

shushanhf commented on src/coreclr/vm/CMakeLists.txt:
```diff
@@ -860,7 +868,22 @@ elseif(CLR_CMAKE_TARGET_ARCH_ARM64)
         )
     endif(CLR_CMAKE_HOST_UNIX)
-elseif(CLR_CMAKE_TARGET_ARCH_LOONGARCH64)
+elseif(clr_cmake_target_arch_loongarch64)
+    set(vm_sources_dac_and_wks_arch
+        ${arch_sources_dir}/stubs.cpp
+        exceptionhandling.cpp
+    )
+
+    set(vm_headers_dac_and_wks_arch
+        ${arch_sources_dir}/virtualcallstubcpu.hpp
+        exceptionhandling.h
+    )
+
+    set(vm_sources_wks_arch
+        ${arch_sources_dir}/profiler.cpp
+        gcinfodecoder.cpp
+    )
```
@clamp03  why did you destroy the LoongArch64's config?????

clamp03 replied:
@shushanhf  I am really sorry. It is just my mistake. I don't want to destroy LoongArch64 at all. Sorry again.

shushanhf replied:
> @shushanhf I am really sorry. It is just my mistake. I don't want to destroy LoongArch64 at all. Sorry again.

shake hands.
